File: src/omni_control/omni_control/kinematics.py
import math
import numpy as np


WHEEL_ORDER = (
    'omni_wheel_joint_1', 'omni_wheel_joint_2',
    'omni_wheel_joint_3', 'omni_wheel_joint_4',
)


def validate_twist(vx_mps, vy_mps, wz_rad_s, max_linear_speed_mps=None,
                   max_angular_speed_rad_s=None):
    """Validate twist values for finiteness and optional physical velocity limits."""
    values = (vx_mps, vy_mps, wz_rad_s)
    if not all(math.isfinite(value) for value in values):
        raise ValueError('twist must contain finite values')
    if max_linear_speed_mps is not None:
        if abs(vx_mps) > max_linear_speed_mps or abs(vy_mps) > max_linear_speed_mps:
            raise ValueError('linear velocity exceeds maximum allowed speed')
    if max_angular_speed_rad_s is not None:
        if abs(wz_rad_s) > max_angular_speed_rad_s:
            raise ValueError('angular velocity exceeds maximum allowed speed')


def inverse_kinematics(vx_mps, vy_mps, wz_rad_s, wheel_radius_m,
                       wheelbase_m, track_width_m, max_wheel_speed_rad_s,
                       wheel_radius_correction=None, kr=None,
                       wheel_radii=None):
    validate_twist(vx_mps, vy_mps, wz_rad_s)
    if not (math.isfinite(wheelbase_m) and math.isfinite(track_width_m)):
        raise ValueError('wheelbase and track width must be finite numbers')
    if wheelbase_m <= 0 or track_width_m <= 0:
        raise ValueError('wheel geometry must be positive')
    if not math.isfinite(max_wheel_speed_rad_s) or max_wheel_speed_rad_s <= 0:
        raise ValueError('max wheel speed must be positive')

    effective_radii = _parse_wheel_radius_correction(
        wheel_radius_m,
        wheel_radius_correction=wheel_radius_correction,
        kr=kr,
        wheel_radii=wheel_radii,
    )

    d = np.sqrt(0.5)
    radius = 0.5 * np.hypot(wheelbase_m, track_width_m)
    # Wheel 1 (FR: -45 deg), Wheel 2 (FL: +45 deg), Wheel 3 (RL: +135 deg), Wheel 4 (RR: -135 deg)
    matrix = np.array([
        [ d,  d, radius],
        [-d,  d, radius],
        [-d, -d, radius],
        [ d, -d, radius],
    ])
    wheel_speeds = (matrix @ np.array([vx_mps, vy_mps, wz_rad_s])) / effective_radii
    scale = max(1.0, float(np.max(np.abs(wheel_speeds))) / max_wheel_speed_rad_s)
    return tuple((wheel_speeds / scale).tolist())


def _parse_wheel_radius_correction(wheel_radius_m, wheel_radius_correction=None,
                                   kr=None, wheel_radii=None):
    if not math.isfinite(wheel_radius_m) or wheel_radius_m <= 0.0:
        raise ValueError('wheel radius must be a finite positive number')

    if wheel_radii is not None:
        arr = np.asarray(wheel_radii, dtype=float)
        if arr.shape != (4,):
            raise ValueError('wheel_radii must be a sequence of 4 elements')
        if not np.all(np.isfinite(arr)):
            raise ValueError('wheel_radii must contain finite values')
        if np.any(arr <= 0.0):
            raise ValueError('wheel_radii elements must be strictly positive')
        return arr

    correction = wheel_radius_correction if wheel_radius_correction is not None else kr
    if correction is None:
        return np.full(4, float(wheel_radius_m), dtype=float)

    arr = np.asarray(correction, dtype=float)
    if arr.ndim == 2:
        if arr.shape != (4, 4):
            raise ValueError('wheel_radius_correction matrix must have shape (4, 4)')
        diag_arr = np.diag(arr)
        if not np.allclose(arr, np.diag(diag_arr)):
            raise ValueError('wheel_radius_correction matrix must be diagonal')
        arr = diag_arr
    elif arr.ndim == 1:
        if arr.shape != (4,):
            raise ValueError('wheel_radius_correction sequence must have length 4')
    else:
        raise ValueError('wheel_radius_correction must be a 4-element sequence or (4, 4) matrix')

    if not np.all(np.isfinite(arr)):
        raise ValueError('wheel_radius_correction must contain finite values')
    if np.any(arr <= 0.0):
        raise ValueError('wheel_radius_correction elements must be strictly positive')

    return float(wheel_radius_m) * arr


def forward_kinematics(wheel_speeds_rad_s, wheel_radius_m, wheelbase_m,
                       track_width_m, wheel_radius_correction=None,
                       kr=None, wheel_radii=None):
    if len(wheel_speeds_rad_s) != 4:
        raise ValueError('four wheel speeds are required')
    speeds = np.array(wheel_speeds_rad_s, dtype=float)
    if not np.all(np.isfinite(speeds)):
        raise ValueError('wheel speeds must be finite')
    if not (math.isfinite(wheelbase_m) and math.isfinite(track_width_m)):
        raise ValueError('wheelbase and track width must be finite numbers')
    if wheelbase_m <= 0 or track_width_m <= 0:
        raise ValueError('wheel geometry must be positive')

    effective_radii = _parse_wheel_radius_correction(
        wheel_radius_m,
        wheel_radius_correction=wheel_radius_correction,
        kr=kr,
        wheel_radii=wheel_radii,
    )
    rim_speeds = speeds * effective_radii

    d = np.sqrt(0.5)
    radius = 0.5 * np.hypot(wheelbase_m, track_width_m)
    pinv_matrix = 0.25 * np.array([
        [ 1.0 / d, -1.0 / d, -1.0 / d,  1.0 / d],
        [ 1.0 / d,  1.0 / d, -1.0 / d, -1.0 / d],
        [ 1.0 / radius, 1.0 / radius, 1.0 / radius, 1.0 / radius],
    ])
    twist = pinv_matrix @ rim_speeds
    return float(twist[0]), float(twist[1]), float(twist[2])


compute_wheel_speeds = inverse_kinematics
compute_body_twist = forward_kinematics

File: firmware/stm32_f407vg_arduino_sim/src/kinematics.cpp
#include "kinematics.h"

#include <math.h>

namespace {

const float kDiagonal = 0.70710678118F;
const float kPi = 3.14159265359F;

bool is_finite(float value) {
    return isfinite(value) != 0;
}

float clamp(float value, float low, float high) {
    return value < low ? low : (value > high ? high : value);
}

}  // namespace

bool compute_wheel_speeds(const TwistCommand &twist,
                          const FirmwareSettings &settings,
                          float wheel_speeds[kWheelCount]) {
    if (settings.wheel_radius_m <= 0.0F ||
        settings.wheelbase_m <= 0.0F ||
        settings.track_width_m <= 0.0F ||
        settings.max_wheel_speed_rad_s <= 0.0F) {
        return false;
    }
    if (!is_finite(twist.vx_mps) || !is_finite(twist.vy_mps) ||
        !is_finite(twist.wz_rad_s)) {
        return false;
    }

    const float radius = 0.5F *
        sqrtf(settings.wheelbase_m * settings.wheelbase_m +
              settings.track_width_m * settings.track_width_m);
    const float values[kWheelCount] = {
        kDiagonal * twist.vx_mps + kDiagonal * twist.vy_mps +
            radius * twist.wz_rad_s,
        -kDiagonal * twist.vx_mps + kDiagonal * twist.vy_mps +
            radius * twist.wz_rad_s,
        -kDiagonal * twist.vx_mps - kDiagonal * twist.vy_mps +
            radius * twist.wz_rad_s,
        kDiagonal * twist.vx_mps - kDiagonal * twist.vy_mps +
            radius * twist.wz_rad_s,
    };
    float scale = 1.0F;
    for (uint8_t index = 0U; index < kWheelCount; ++index) {
        wheel_speeds[index] = values[index] / settings.wheel_radius_m;
        const float magnitude = fabsf(wheel_speeds[index]);
        if (magnitude > settings.max_wheel_speed_rad_s) {
            scale = fmaxf(scale, magnitude / settings.max_wheel_speed_rad_s);
        }
    }
    if (scale > 1.0F) {
        for (uint8_t index = 0U; index < kWheelCount; ++index) {
            wheel_speeds[index] /= scale;
        }
    }
    return true;
}

bool compute_body_twist(const float wheel_speeds[kWheelCount],
                        const FirmwareSettings &settings,
                        TwistCommand &twist) {
    if (settings.wheel_radius_m <= 0.0F ||
        settings.wheelbase_m <= 0.0F ||
        settings.track_width_m <= 0.0F) {
        return false;
    }
    for (uint8_t index = 0U; index < kWheelCount; ++index) {
        if (!is_finite(wheel_speeds[index])) {
            return false;
        }
    }
    const float radius = 0.5F *
        sqrtf(settings.wheelbase_m * settings.wheelbase_m +
              settings.track_width_m * settings.track_width_m);
    twist.vx_mps = settings.wheel_radius_m *
        (wheel_speeds[0] - wheel_speeds[1] - wheel_speeds[2] +
         wheel_speeds[3]) / (4.0F * kDiagonal);
    twist.vy_mps = settings.wheel_radius_m *
        (wheel_speeds[0] + wheel_speeds[1] - wheel_speeds[2] -
         wheel_speeds[3]) / (4.0F * kDiagonal);
    twist.wz_rad_s = settings.wheel_radius_m *
        (wheel_speeds[0] + wheel_speeds[1] + wheel_speeds[2] +
         wheel_speeds[3]) / (4.0F * radius);
    return true;
}

float wrap_angle(float angle_rad) {
    while (angle_rad > kPi) {
        angle_rad -= 2.0F * kPi;
    }
    while (angle_rad < -kPi) {
        angle_rad += 2.0F * kPi;
    }
    return angle_rad;
}